File: cpp/FoodChain/FoodChain/UnionFindTree.cpp
//
//  UnionFindTree.cpp
//  FoodChain
//
//  Created by 岩佐 淳史 on 2013/05/17.
//  Copyright (c) 2013年 岩佐 淳史. All rights reserved.
//

#include "UnionFindTree.h"

UnionFindTree::UnionFindTree(int size) : data(size, -1)
{
}

bool UnionFindTree::unionSet(int x, int y) {
    x = root(x); y = root(y);
    if (x != y) {
        if (data[y] < data[x]) std::swap(x, y);
        data[x] += data[y]; data[y] = x;
    }
    return x != y;
}

bool UnionFindTree::findSet(int x, int y) {
    return root(x) == root(y);
}

int UnionFindTree::root(int x) {
    return data[x] < 0 ? x : data[x] = root(data[x]);
}

int UnionFindTree::size(int x) {
    return -data[root(x)];
}

File: java/MilitaryDraft/src/main/java/com/bloodysnow/UnionFindTree.java
package com.bloodysnow;

import java.util.ArrayList;
import java.util.List;

/**
 * Created with IntelliJ IDEA.
 * User: Atsushi
 * Date: 2013/07/05
 * Time: 11:05
 * To change this template use File | Settings | File Templates.
 */
public class UnionFindTree {
    private int[] data;
    public UnionFindTree(int size) {
        data = new int[size];
        for(int i = 0; i < size; ++i) data[i] = -1;
    }
    public boolean unionSet(int x, int y) {
        x = root(x);
        y = root(y);
        if(x != y) {
            if (data[y] < data[x]) {
                int tmp = data[x];
                data[x] = data[y];
                data[y] = tmp;
            }
            data[x] += data[y];
            data[y] = x;
        }
        return x != y;
    }
    public boolean findSet(int x, int y) { return root(x) == root(y); }
    public int root(int x) {
        if(data[x] < 0) return x;
        else return data[x] = root(data[x]);
    }
    public int size(int x) { return -data[root(x)]; }
}
